### ISSUE TITLE
Massive regression in download speed in Seafile Windows client 9.0.7

### ISSUE THREAD

**Issue by Makumasa** [haiwen/seafile#2817]
I use the Seafile desktop client on multiple machines on my home network, which I use to sync files to/from a Seafile server also running on my home network. Recently, I've noticed that the download speed in the client has been very slow. For example, on my PC, my download speeds would typically range between 100-140 MB/s, but they were now 10-30 MB/s. Running a speed test confirmed that my machine's network performance hadn't changed, and downloading the same file via the Seafile web UI was substantially faster, indicating that the client was the most likely culprit. In my attempts to troubleshoot this issue, I tried the following:

- Using a different machine
- Downloading one large file
- Downloading many small files
- Downloading from an unencrypted library
- Downloading using a different user
- Upgrading the client to 9.0.8

None of these resolved the issue. No matter what I tried, the download speed would never exceed 30 MB/s, and would typically hover around 17 MB/s. My brother's laptop, however, was getting download speeds above 100 MB/s on the same network. After discovering that his client was still on version 9.0.6, I tried rolling back my machines to 9.0.6, and sure enough the throughput returned to what I'd expect.

I tried looking through the client logs, but didn't find anything out of the ordinary, so I'm not sure what the root cause is. The most I can say with any confidence is that the regression appears to have been introduced between versions 9.0.6 and 9.0.7 of the Seafile Windows client. I am unsure if the Linux/Mac clients are impacted, as I haven't tested those.

**Comment by killing** [comment]
In 9.0.7 version we modified the download process to directly write the downloaded data into the final file. In previous versions the data was written first into a temp file in internal cache folder then write to the user-visible file. We'll analyze the code for performance issues.

**Comment by Makumasa** [comment]
> In 9.0.7 version we modified the download process to directly write the downloaded data into the final file. In previous versions the data was written first into a temp file in internal cache folder then write to the user-visible file. The purpose of this change is to reduce the potential cache size limitation when downloading large files. We'll analyze the code for performance issues.

Sounds good. Let me know if you'd like me to provide any additional information from my client or the tests I ran. As far as I can tell, the issue is trivial to reproduce.

**Comment by killing** [comment]
This is fixed in 9.0.9. Please test it again. Note that the speed for encrypted library is not improved, to avoid over-complicated logic.

I'll leave this issue open as encrypted library is not fixed.

**Comment by Makumasa** [comment]
Happy to hear it! I use encrypted libraries, so I'll hold off updating until that's addressed.

**Comment by Makumasa** [comment]
Alright, just tested on 9.0.9 on Windows and I can confirm that the download speed regression seems to be fixed for unencrypted libraries, but not for encrypted libraries.

Since I use encrypted libraries, I'll hold off on updating my systems until the issue is fixed for encrypted libraries, as well. Thanks again!

**Comment by Makumasa** [comment]
Hello, just checking to see if there's any updates on addressing this issue for encrypted libraries. I noticed the issue was tagged with priority-low, which is a bit surprising given that the issue causes a 10x or more reduction in download speed. Are encrypted libraries just not widely used?

**Comment by killing** [comment]
No, currently the issue for encrypted libraries is not addressed yet. We'll look into improvements in future versions. I've also bumped up the priority.

**Comment by feiniks** [comment]
Hello,

This issue has been fixed in version 9.0.21. You can upgrade to version 9.0.21 and check whether the issue has been resolved.

**Comment by Makumasa** [comment]
> Hello,
> 
> This issue has been fixed in version 9.0.21. You can upgrade to version 9.0.21 and check whether the issue has been resolved.

Yeah, I noticed that the changelog mentioned a resolution to the encrypted library download performance issue. I've updated my machines to 9.0.21 and can confirm that the issue is resolved (I uploaded a large video file to an encrypted library via the web UI and monitored the download speed on the desktop client).

That said, it appears that updating from 9.0.6 to 9.0.21 broke my ability to sync directory junctions. I did a cursory search for issues related to "junctions" on both this repo and the `seafile-client` repo and didn't find anything - is this a known issue, or should I create a new issue for it?
